### PR TITLE
[CUDA] Use accumulate type to improve accuracy of grid_sample on half precision inputs

### LINKED ISSUE #89836: F.grid_sampling with half precision type has very large numeric error
### 🐛 Describe the bug

when use torch.float16 data type as input to F.grid_sampling, output has very large gap compared with torch.float64 inputs.

the demo code outputs:
```
max_abs_err: 0.02377164; max_rel_err: 223.61935484; mean_rel_err: 0.02276072; 
```

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

torch.manual_seed(0)

@torch.no_grad()
def generate_inputs_grid_sampling(dtype):
    N, C, Hi, Wi = 32, 64, 16, 16
    N, Ho, Wo = N, Hi//2, Wi//2
    
    inputs = torch.randn([N, C, Hi, Wi])
    grid = torch.rand([N, Ho, Wo, 2]) * 2 - 1
    return (
        inputs.cuda().to(dtype),
        grid.cuda().to(dtype),
    )

def calc_err(x1, x2):
    err = (x1 - x2).abs()
    err_rel = err / x2.abs()
    max_abs_err = err.max()
    max_rel_err = err_rel.max()
    mean_rel_err = err_rel.mean()
    return max_abs_err.item(), max_rel_err.item(), mean_rel_err.item()

def sampling(inputs, grid):
    return F.grid_sample(inputs, grid, mode='bilinear', padding_mode='zeros', align_corners=False)

def info_versions():
    print(f'Pytorch version: {torch.__version__}')
    print(f'CUDA version: {torch.version.cuda}')
    print(f'NCCL version: {torch.cuda.nccl.version()}')

if __name__ == '__main__':
    info_versions()

    inputs, grid = generate_inputs_grid_sampling(dtype=torch.float16)
    out = sampling(inputs, grid)

    inputs_fp64, grid_fp64 = inputs.double(), grid.double()
    out_fp64 = sampling(inputs_fp64, grid_fp64)

    max_abs_err, max_rel_err, mean_rel_err = calc_err(out, out_fp64)

    print(f'max_abs_err: {max_abs_err:.8f}; max_rel_err: {max_rel_err:.8f}; mean_rel_err: {mean_rel_err:.8f}; ')

```

### Versions

```
Collecting environment information...
PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.27

Python version: 3.9.13 (main, Oct 13 2022, 21:15:33)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-131-generic-x86_64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1060 6GB
Nvidia driver version: 470.141.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.3
[pip3] torch==1.12.1
[pip3] torch-ort==1.12.0
[pip3] torchaudio==0.12.1
[pip3] torchvision==0.13.1
[conda] blas                      1.0                         mkl    http://maven.opencloud.sensetime.com/repository/conda-tsinghua-main
[conda] cudatoolkit               11.3.1               h2bc3f7f_2    http://maven.opencloud.sensetime.com/repository/conda-tsinghua-main
[conda] ffmpeg                    4.3                  hf484d3e_0    http://maven.opencloud.sensetime.com/repository/conda-tsinghua-cloud/pytorch
[conda] mkl                       2021.4.0           h06a4308_640    http://maven.opencloud.sensetime.com/repository/conda-tsinghua-main
[conda] mkl-service               2.4.0            py39h7f8727e_0    http://maven.opencloud.sensetime.com/repository/conda-tsinghua-main
[conda] mkl_fft                   1.3.1            py39hd3c417c_0    http://maven.opencloud.sensetime.com/repository/conda-tsinghua-main
[conda] mkl_random                1.2.2            py39h51133e4_0    http://maven.opencloud.sensetime.com/repository/conda-tsinghua-main
[conda] numpy                     1.23.3           py39h14f4228_0    http://maven.opencloud.sensetime.com/repository/conda-tsinghua-main
[conda] numpy-base                1.23.3           py39h31eccc5_0    http://maven.opencloud.sensetime.com/repository/conda-tsinghua-main
[conda] pytorch                   1.12.1          py3.9_cuda11.3_cudnn8.3.2_0    http://maven.opencloud.sensetime.com/repository/conda-tsinghua-cloud/pytorch
[conda] pytorch-mutex             1.0                        cuda    http://maven.opencloud.sensetime.com/repository/conda-tsinghua-cloud/pytorch
[conda] torch-ort                 1.12.0                   pypi_0    pypi
[conda] torchaudio                0.12.1               py39_cu113    http://maven.opencloud.sensetime.com/repository/conda-tsinghua-cloud/pytorch
[conda] torchvision               0.13.1               py39_cu113    http://maven.opencloud.sensetime.com/repository/conda-tsinghua-cloud/pytorch
```

### PR DESCRIPTION
**Pull request by xwang233**
Fixes https://github.com/pytorch/pytorch/issues/89836

This PR changes the CUDA kernels of grid_sample 2d and 3d, forward, to use accumulate type to improve accuracy on half precision inputs.

Also, the backward error on grad with half input is in the order of 1e-4, unlike 1e2 in forward process. The backward kernels are thus unchanged.

### CODE REVIEW

**Comment by xwang233** [comment]
Hi @ngimel , the CI looks green for all the tests now.

File: aten/src/ATen/native/cuda/GridSampler.cu [hunk 1]
```diff
@@ -754,9 +757,10 @@ void launch_grid_sampler_2d_forward_kernel(
   int64_t count = N * H * W;
   if (count > 0) {
     AT_DISPATCH_FLOATING_TYPES_AND_HALF(input.scalar_type(), "grid_sampler_2d_cuda", [&] {
+      using accscalar_t = at::acc_type<scalar_t, true>;
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/GridSampler.cu:760]: can you please use opmath_t here? Also, no need for extra template parameter, you can do `using opmath_t = ...` directly in the kernel (although it doesn't really matter, number of instantiations is still the same).

**Review by ngimel — COMMENTED** [review]
Minor comment, otherwise looks good. I'm surprised backward error is not large, but anyway, not a lot of choice we have there, as gradients are atomically accumulated in the outputs.

File: aten/src/ATen/native/cuda/GridSampler.cu [hunk 2]
```diff
@@ -57,8 +60,8 @@ namespace {
       scalar_t x = grid.data[grid_offset];
       scalar_t y = grid.data[grid_offset + grid_sCoor];
 
-      scalar_t ix = grid_sampler_compute_source_index(x, inp_W, padding_mode, align_corners);
-      scalar_t iy = grid_sampler_compute_source_index(y, inp_H, padding_mode, align_corners);
+      opmath_t ix = grid_sampler_compute_source_index(x, inp_W, padding_mode, align_corners);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/GridSampler.cu:63]: you have to cast `x` to `opmath_t` if you want higher precision result, otherwise `grid_sampler_compute_source_index` would run in `scalar_t`, and just upcast low-precision result.

Comment (xwang233) [reply aten/src/ATen/native/cuda/GridSampler.cu:63]: Sure. I have tried to promote `x`  to `opmath_t` on  `grid_sampler_compute_source_index`, but that would cause unit test failure on `test_grid_sample_large, issue_24823_1, torch.half`.

Specifically, in that test, `affine_grid` was done in half precision and created an output of (half) `x = 0.666503906250`. By calculating
```
((coord + 1.f) * size - 1) / 2
```
of `grid_sampler_unnormalize` in half precision we get exactly 2.0, but 1.999755859375 in float precision. With the latter value, it falls "within in bound" of [0, 2) and yields wrong results.

Since using non-promoted `scalar_t` dtype in `grid_sampler_compute_source_index` still yields the accurate result, as shown in the added unit test with half precision, I think it's ok to use the non-promoted type in this step. A more comprehensive solution would be to use promoted type in `affine_grid` as well, but that would take extra effort and beyond the scope of this PR's fix. What do you think?

Here is a short CUDA program that demonstrates this
```cuda
#include <cuda.h>
#include <cuda_runtime.h>
#include <cuda_fp16.h>

#include <iostream>

enum class GridSamplerPadding {Zeros, Border, Reflection};

template <typename scalar_t>
static __forceinline__ __device__
scalar_t grid_sampler_unnormalize(scalar_t coord, int size, bool align_corners) {
  // unnormalize coord from [-1, 1] to [-0.5, size - 0.5]
  return (((float)coord + 1.f) * size - 1) / 2;
}

// Computes the pixel source index value for a grid coordinate
template <typename scalar_t>
static __forceinline__ __device__
scalar_t grid_sampler_compute_source_index(
    scalar_t coord,
    int size,
    GridSamplerPadding padding_mode,
    bool align_corners) {
  printf("coord 1 = %.12f\n", (float)coord);
  coord = grid_sampler_unnormalize(coord, size, align_corners);
  printf("coord 2 = %.12f\n", (float)coord);
  return coord;
}

template <typename scalar_t>
__global__ void kernel(scalar_t* coord) {
  int size = 3;
  GridSamplerPadding padding_mode = GridSamplerPadding::Zeros;
  bool align_corners = false;

  scalar_t coord_out = grid_sampler_compute_source_index(*coord, size, padding_mode, align_corners);
  printf("coord_out = %.12f\n", (float)coord_out);
}

#ifndef TYPE_T
  #define TYPE_T float
#endif

int main() {
  // using type_t = __half;
  // using type_t = float;
  using type_t = TYPE_T;

  type_t h_coord = 0.666503906250;
  type_t *d_coord;
  cudaMalloc(&d_coord, sizeof(type_t) * 1);
  cudaMemcpy(d_coord, &h_coord, sizeof(type_t) * 1, cudaMemcpyHostToDevice);

  kernel<<<1, 1>>>(d_coord);
  cudaDeviceSynchronize();
}
```

Comment (ngimel) [reply aten/src/ATen/native/cuda/GridSampler.cu:63]: So in this particular case lower precision actually passes the test due to some compounding of errors? I'm pretty sure it will be different for other inputs, but ok we can leave it like this to unblock

**Review by ngimel — APPROVED** [review]
